File: 500/503 - Next Greater Element II.java
import java.util.Stack;

class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int[] ret = new int[nums.length];
        Stack<Integer> stack = new Stack<Integer>();
        int i, num;
        
        for (i = 0; i < nums.length; ++i)
            ret[i] = -1;
        
        for (i = 0; i < 2 * nums.length; ++i) {
            num = nums[i % nums.length];
            
            while ((! stack.isEmpty()) && nums[stack.peek()] < num) 
                ret[stack.pop()] = num;
            
            if (i < nums.length)
                stack.push(i);
        }
        
        return ret;
    }
}

File: csharp/503.next-greater-element-ii.cs
/*
 * @lc app=leetcode id=503 lang=csharp
 *
 * [503] Next Greater Element II
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {
    public int[] NextGreaterElements(int[] nums) {
        int n = nums.Length, m = 2 * n;
        IList<int> stack = new List<int>();
        int[] ret = new int[n];
        for (int i = 0; i < m; ++i) {
            int num;
            if (i < n) {
                ret[i] = -1;
                num = nums[i];
            } else {
                num = nums[i - n];
            }
            while (stack.Count > 0 && nums[stack[stack.Count - 1]] < num) {
                ret[stack[stack.Count - 1]] = num;
                stack.RemoveAt(stack.Count - 1);
            }
            if (i < n) {
                stack.Add(i);
            }
        }
        return ret;
    }
}
// @lc code=end

